keps/sig-node/4680-add-resource-health-to-pod-status/README.md

```diff
@@ -285,53 +285,113 @@ We should consider introducing another field to the Status that will be a free f
+        *   `getHealthInfo(driverName, poolName, deviceName)`: Returns current health, considering `healthTimeout`. If `time.Since(device.LastUpdated) > healthTimeout`, it returns "Unknown".
+        *   `clearDriver(driverName)`: Removes all data for a driver (e.g., on deregistration or stream end).
+        *   `saveToCheckpoint()` / `loadFromCheckpoint()`: Handles persistence via JSON marshalling.
+        *   `healthTimeout`: A constant (e.g., 30s) for marking stale devices "Unknown".
+    *   **Handling Timeouts and 'Unknown' State:**
+        *   "Unknown" status due to timeout is determined on-demand by `getHealthInfo`.
+        *   `updateHealthInfo` also actively sets "Unknown" for devices previously known but not in the latest update from the plugin, if their `LastUpdated` timestamp exceeds `healthTimeout`.
+        *   No separate background expiration worker is required; Kubelet's existing status manager update triggers will naturally reflect timeouts when `getHealthInfo` is called.
+        *   If a plugin stops sending updates for all its devices (e.g., plugin crashes, stream ends), `HandleWatchResourcesStream` calls `clearDriver`, and subsequent calls to `getHealthInfo` for these devices will return "Unknown".
+
+2.  **gRPC API for DRA Device Health (`staging/src/k8s.io/kubelet/pkg/apis/dra-health/v1alpha1/api.proto`)**
+    *   A new API group `dra-health/v1alpha1` is introduced.
+    *   Defines a `NodeHealth` service:
+        ```proto
+        service NodeHealth {
+          // WatchResources allows a DRA plugin to stream health updates for its devices to Kubelet.
+          // Kubelet calls this method, and the plugin streams responses.
+          // This method is optional; if not implemented by a plugin, Kubelet will assume
+          // devices managed by that plugin have an "Unknown" health status.
+          rpc WatchResources(WatchResourcesRequest) returns (stream WatchResourcesResponse) {}
+        }
+
+        message WatchResourcesRequest {
+          // Reserved for future use, e.g., filtering or options.
+        }
+
+        message WatchResourcesResponse {
+          // A list of all devices managed by the plugin for which health is being reported.
+          // This should be a complete list for the driver; Kubelet will reconcile this state.
+          repeated DeviceHealth devices = 1;
+        }
+
+        message DeviceHealth {
+          // The name of the resource pool this device belongs to.
+          // Required.
+          string pool_name = 1;
+          // The unique name of the device within the pool.
+          // Required.
+          string device_name = 2;
```

Comment:
xref: https://github.com/kubernetes/enhancements/pull/5104#discussion_r2101136766

TLDR: "device name" will become `<device name>[/<share ID>]` once consumable capacities allow allocating the same device multiple times. This is consistent with the proposed API here, it just needs an updated description - but not now! It depends on the order in which the implementations get merged.

cc @sunya-ch